### ISSUE TITLE
Importance sampling

### ISSUE THREAD

**Issue by SushantDaga** [brendanhogan/DeepSeekRL-Extended#5]
https://github.com/brendanhogan/DeepSeekRL-Extended/blob/09e312a07f634cbb6ee4aaf2d9f3372dca9e7b9b/main.py#L315

Hey, shouldn't above be

```python
importance_ratio = torch.exp(per_token_logps - per_token_logps_ref.detach()) # subtracted term is **per_token_logps_ref**
per_token_loss =  importance_ratio * advantages.unsqueeze(1) 
```

Right now ratio is amounting to 1.0 always [ $e^{(x-x)} = e^{0} \equiv 1.0$ ]

PS: thanks for this repo. It is really useful in implementing GRPO from scratch :)

**Comment by brendanhogan** [comment]
great question! And its something that through me off too - but if you look at TRL's implementation: 
https://github.com/huggingface/trl/blob/main/trl/trainer/grpo_trainer.py

in the function ```def _compute_loss(self, model, inputs):```
They actually do the same and say: 
```
        # When using num_iterations == 1 and steps_per_generation <= gradient_accumulation_steps
        # old_per_token_logps == per_token_logps, so we can skip it's computation
        # (see _generate_and_score_completions) and use per_token_logps.detach() instead.
        old_per_token_logps = (
            per_token_logps.detach() if inputs["old_per_token_logps"] is None else inputs["old_per_token_logps"]
        )
        coef_1 = torch.exp(per_token_logps - old_per_token_logps)
```

So I think if you are doing like old batches, or some kind of replay buffer it does need to be different  - but assuming the same model you are training did the generations its actually fine that its just the advantage as the loss

And thank you! I hope it is easy to follow! I think the later branches get a little confusing - I need to clean them up
